bdai_ros2_wrappers/bdai_ros2_wrappers/subscription.py

```diff
@@ -78,3 +81,137 @@ def wait_for_message(
         future.cancel()
         return None
     return future.result()
+
+
+def wait_for_messages(
+    node: rclpy.node.Node,
+    topics: typing.List,
+    mtypes: typing.List,
+    **kwargs: typing.Any,
+) -> typing.Any:
+    """Waits for messages to arrive at multiple topics within a given time window.
+
+    Uses message_filters.ApproximateTimeSynchronizer. This function blocks
+    until receiving the messages or when a given timeout expires. Assumes the
+    given node is spinning by some external executor.
+
+    Requires the user to pass in a node, since
+    message_filters.Subscriber requires a node upon construction.
+
+    Args:
+        node (Node): the node being attached
+        topics (list): List of topics
+        mtypes (list): List of message types, one for each topic.
+        kwargs: additional arguments, including
+            delay  (float) The delay in seconds for which the messages
+                could be synchronized.
+            allow_headerless (bool): Whether it's ok for there to be
+                no header in the messages.
+            sleep (float) the amount of time to wait before checking
+                whether messages are received
+            timeout (float or None): Time in seconds to wait. None if forever.
+                If exceeded timeout, self.messages will contain None for
+                each topic.
+            latched_topics (set): a set of topics for which the publisher latches (i.e.
+                sets QoS durability to transient_local).
+
+    """
+    return _WaitForMessages(node, topics, mtypes, **kwargs).messages
+
+
+class _WaitForMessages:
+    def __init__(
```

Comment:
@kaiyu-zheng meta: using `__init__` methods to do work is a fairly unusual idiom. If you want to bundle the logic in an object for reuse (rather than having a temporary instantiated on each `wait_for_messages` invocation), consider relocating to a e.g. `wait_for_messages()` method.

Reply:
I agree that this is unusual.  I could refactor so that the logic in `__init__` is in another method in `_WaitForMessages`. If we do the `futures` route as proposed above, then that would affect this refactoring I suppose. But that also influences user experience (and the test). I think we should first agree about that.

Reply:
Again thanks for the advice. Refactored!

(Ignore my comment above)